src/eval.rs

```diff
@@ -0,0 +1,216 @@
+use crate::parse::BinOpKind;
+use crate::scope::*;
+use crate::value::*;
+use crate::EvalError;
+use gc::{Finalize, Gc, GcCell, Trace};
+use rnix::TextRange;
+use std::borrow::Borrow;
+
+type TreeResult = Result<Box<Tree>, EvalError>;
+
+/// Used to lazily calculate the value of a Tree. This should be
+/// tolerant of parsing and evaluation errors from child Trees.
+#[derive(Debug, Clone, Trace, Finalize)]
+pub enum TreeSource {
+    Literal {
+        value: NixValue,
+    },
+    Paren {
+        inner: TreeResult,
+    },
+    BinOp {
+        op: BinOpKind,
+        left: TreeResult,
+        right: TreeResult,
+    },
+    BoolAnd {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    BoolOr {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    Implication {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    UnaryInvert {
+        value: TreeResult,
+    },
+    UnaryNegate {
+        value: TreeResult,
+    },
+}
+
+/// Syntax node that has context and can be lazily evaluated.
+#[derive(Clone, Trace, Finalize)]
+pub struct Tree {
```

Comment:
Wouldn't it make sense to call this e.g. `TreeNode` or `ExprNode`?

Reply:
Would `Expr` be a good name?

Reply:
I guess so.